[HIP-0026] Plugin packaging, signing, and verification

scottrigby opened the pull request:
Adds plugin packaging, signing, and verification commands.

Part of implementing hip-0026. For full context, see:
- https://github.com/helm/helm/issues/31144

Depends on:
- #31146 
- #31172 
- #31174 (for now this PR includes the commit from there cherry-picked)

Please review these PRs first. After these PRs are merged, will change this PR base branch to `main` and move out of draft.

Checklist:
- [x] This PR contains user facing changes
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

## Manual testing

you should be able to verify testing the tarballs in https://github.com/scottrigby/h4-example-plugins/tree/main/tarballs using my public key. You can either use the keybase method as described here https://helm.sh/docs/topics/provenance/#verifying-packages, or you can also just download my public key from here https://keybase.io/r6by/pgp_keys.asc and import with gpg.

You can of course also use your own key to test packging with signing and verifying the packages you made.

## Changes in this PR

1. adds `plugin package` and `plugin verify` commands

    attempts to sign the package by default. This plugin command has the same CLI options as `helm package`. You may use `--sign=false` to skip signing (not recommended for distribution, and emits a warning)

1. adds `PROVENANCE` status to `plugin list` output

    ```
    % go run ./cmd/helm plugin list
    NAME                     	VERSION	TYPE     	APIVERSION	PROVENANCE	SOURCE                                          
    example-cli              	0.1.0  	cli/v1   	v1        	signed    	https://github.com/scottrigby/h4-example-plugins
    example-getter           	0.1.0  	getter/v1	v1        	signed    	https://github.com/scottrigby/h4-example-plugins
    example-legacy-cli       	0.1.0  	cli/v1   	legacy    	signed    	unknown                                         
    example-legacy-downloader	0.1.0  	getter/v1	legacy    	signed    	unknown                                         
    example-unsigned         	0.1.0  	cli/v1   	v1        	unsigned  	https://github.com/scottrigby/h4-example-plugins
    ```
    provenance status can be:
    - "local dev": Plugin is a symlink (development mode)
    - "unsigned": No provenance file found
    - "invalid provenance": Provenance file is malformed
    - "mismatched provenance": Provenance file does not exactly match plugin
    - "signed": Valid signature exists for this exact plugin
1. `plugin install` now attempts to verify provenance by default:
    ```
    plugin install path/to/example-cli-0.1.0.tgz
    Verifying plugin signature...
    Signed by: Scott Rigby <scott@r6by.com>
    Using Key With Fingerprint: 208DD36ED5BB3745A16743A4C7C6FBB5B91C1155
    Plugin Hash Verified: sha256:4ef0c4b1cd402d6a170fadb14d68733226e37bff9e2df67ee34754f4a6d0d1c6
    Installed plugin: example-cli
    ```
    it also downloads the tarball along with the prov file alongside the unpacked plugin, inside the HELM_PLUGINS directory.

    if a correctly named provenance file exists next to the plugin — whether local directory, a local or HTTP tarball, or in the same OCI artifact, helm install will attempt verification. If the chart passes verification, the unpacked plugin dir and provenance file will be downloaded to the HELM_PLUGINS directory. Example:

    ```
    % ls -lah $(go run ./cmd/helm env HELM_PLUGINS)                               
    drwxr-xr-x@  4 r6by  staff   128B Aug 25 05:42 example-cli
    -rw-r--r--@  1 r6by  staff   511B Aug 25 05:42 example-cli.tgz
    -rw-r--r--@  1 r6by  staff   1.4K Aug 25 05:42 example-cli.tgz.prov
    ```
    If there is no provenance file, or if verification fails, the plugin will not be downloaded. You may bypass this with `--verify=false`, but the user is warned this is insecure if they do this. It is secure by default.
1. `plugin uninstall` now also removes the plugin's accompanying tarball and provenance file from HELM_PLUGIN directory if they exist locally
1. added `plugin verify`

    this accepts a path to a plugin tarball, which we also save for you in HELM_PLUGINS dir on `plugin install`

scottrigby committed:
wrap keyring read error with more explicit message

Co-authored-by: Jesse Simpson <jesse.simpson36@gmail.com>
Signed-off-by: Scott Rigby <scott@r6by.com>

jessesimpson36 commented on pkg/cmd/plugin_package.go:
```diff
@@ -0,0 +1,206 @@
+	f.StringVar(&o.key, "key", "", "name of the key to use when signing. Used if --sign is true")
+	f.StringVar(&o.keyring, "keyring", defaultKeyring(), "location of a public keyring")
+	f.StringVar(&o.passphraseFile, "passphrase-file", "", "location of a file which contains the passphrase for the signing key. Use \"-\" to read from stdin.")
+	f.StringVarP(&o.destination, "destination", "d", ".", "location to write the plugin tarball.")
```
I think the default being `.` is problematic. on my system, when I run
```
/home/jesse/code/helm/bin/helm plugin package --key "Helm Signing Key" --keyring ./exported-keyring.gpg  ./
```

It saves the plugin package to `..tgz.prov` and `..tgz`

```
Successfully signed. Signature written to: ..tgz.prov
Successfully packaged plugin and saved it to: ..tgz
```

makes it hard to find the package.

jessesimpson36 commented on pkg/cmd/plugin_package.go:
```diff
@@ -0,0 +1,206 @@
+
+	// Now create the tarball (only after signing prerequisites are met)
+	pluginName := filepath.Base(o.pluginPath)
+	tarballPath := filepath.Join(o.destination, pluginName+".tgz")
```
I think pluginName should actually be from the `plugin.yaml` configuration file.

jessesimpson36 reviewed (changes requested):
See above comments. Otherwise, I'm happy.

jessesimpson36 commented on internal/plugin/installer/local_installer.go:
```diff
@@ -62,8 +99,102 @@ func (i *LocalInstaller) Install() error {
+	}
+
+	return helmpath.DataPath("plugins", pluginName)
+}
```
I feel like `Path()` should be the name of the plugin, and not the directory name. Ideally this would be the same, but in my case, I was trying to install `./` , and then it created the file `..tgz`. Which when I then tried to install the plugin from `..tgz`, it failed because `..` exists already within the ~/.cache/helm/plugins dir

scottrigby committed:
Plugin packaging, signing and verification

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby committed:
Fix test isolation: Add missing ensure.HelmHome(t) to TestExtractPluginInSubdirectory

This test was writing to the user's actual Helm data and cache directories

Signed-off-by: Scott Rigby <scott@r6by.com>

banjoh reviewed (commented):
Changes look good. I added a few minor comments

scottrigby committed:
skip unnecessary check

Co-authored-by: Evans Mungai <mbuevans@gmail.com>
Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby commented on internal/plugin/installer/http_installer.go:
```diff
@@ -175,111 +142,62 @@ func (i HTTPInstaller) Path() string {
-	// is not a good practice.
-	if slices.Contains(strings.Split(dest, "/"), "..") {
-		return "", errors.New("path contains '..', which is illegal")
+	cleanup := func() {
```
The cleanup logic is a bit convoluted. The complexity exists because:

1. InstallWithOptions() calls PrepareForVerification() which writes temporary files for HTTP/OCI installers (but not Local)
2. pkg/provenance functions currently require file paths, not data
3. Only some installers need cleanup, hence the optional cleanup function pattern

The root cause is that verification currently requires files on disk. I've created a follow-up PR (#31196) that refactors pkg/provenance to accept data directly instead of file paths, eliminating the need for temporary files and this cleanup complexity entirely.

Let's merge this PR as-is to get the immediate improvements, then address the I/O elimination in the follow-up PR.

scottrigby commented on internal/plugin/verify.go:
```diff
@@ -0,0 +1,72 @@
+		return nil, errors.New("directory verification not supported - only plugin tarballs can be verified")
+	}
+
+	// Verify it's a tarball
+	if !isTarball(pluginPath) {
+		return nil, errors.New("plugin file must be a gzipped tarball (.tar.gz or .tgz)")
+	}
```
You're correct that this is overly restrictive. This function assumes plugin signing is only for tarballs (similar to chart signing), but the logic could be more robust.

However, this logic is entirely removed in the follow-up PR (#31196) as part of eliminating the file-based verification approach. Since it's a minor issue that gets completely resolved in the follow-up, let's merge this PR and address it there.

banjoh reviewed (approved):
LGTM. Ship it!

scottrigby committed:
Fix display signing status

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby commented:
@sabre1041 f4ded70 fixes the issue you found

scottrigby committed:
Fix mismatched provenance

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby commented:
Fixed the "mismatched provenance" status. Now it works as expected: f you change the hash in the provenance file, or the tarball is not exactly what was signed (tampered with, wrong tarball or version, etc), the mismatched status shows correctly.

scottrigby commented:
test statuses like this:

```
# install signed plugin
% go run ./cmd/helm plugin install https://github.com/scottrigby/h4-example-plugins/raw/refs/heads/main/tarballs/example-cli-0.1.0.tgz
Verifying plugin signature...
Signed by: Scott Rigby <scott@r6by.com>
Using Key With Fingerprint: 208DD36ED5BB3745A16743A4C7C6FBB5B91C1155
Plugin Hash Verified: sha256:4ef0c4b1cd402d6a170fadb14d68733226e37bff9e2df67ee34754f4a6d0d1c6
Installed plugin: example-cli

# install unsigned plugin (must pass --verify=false)
% go run ./cmd/helm plugin install --verify=false https://github.com/scottrigby/h4-example-plugins/raw/refs/heads/main/tarballs/example-unsigned-0.1.0.tgz
WARNING: Skipping plugin signature verification
Installed plugin: example-unsigned

# manually install plugin with mismatched prov
% cd $(go run ./cmd/helm env HELM_PLUGINS)
% curl -SsLO https://github.com/scottrigby/h4-example-plugins/raw/refs/heads/main/tarballs/example-mismatched-provenance-0.1.0.tgz
% curl -SsLO https://github.com/scottrigby/h4-example-plugins/raw/refs/heads/main/tarballs/example-mismatched-provenance-0.1.0.tgz.prov
% tar -zxvf example-mismatched-provenance-0.1.0.tgz 
x example-mismatched-provenance
x example-mismatched-provenance/plugin.yaml

# see statuses
% go run ./cmd/helm plugin list
NAME                         	VERSION	TYPE  	APIVERSION	PROVENANCE           	SOURCE                                          
example-cli                  	0.1.0  	cli/v1	v1        	signed               	https://github.com/scottrigby/h4-example-plugins
example-mismatched-provenance	0.1.0  	cli/v1	v1        	mismatched provenance	https://github.com/scottrigby/h4-example-plugins
example-unsigned             	0.1.0  	cli/v1	v1        	unsigned             	https://github.com/scottrigby/h4-example-plugins
```

sabre1041 commented:
> > @scottrigby Performed some additional testing and from an end user experience, defaulting verification to `true` could prove troublesome. The following is an alternate approach that we might be able to consider
> > 
> > 1. `--verify` option false by default if no `.prov` file found. This aligns with the functionality of Helm charts
> 
> ok i can see this. if .prov is not found, do you think it's acceptable from a user experience for helm to emit a warning rather than a failure?
> 
> > 2. Attempt to verify if `.prov` file found. User can set `--verify` option which they can use to disable chart verification if desired
> 
> ok great, that's how this part works now. can update the above if you agree about warning

@scottrigby agree with the approach as it provides a balance of ensuring that existing plugins are unaffected functionally, but illustrates the benefits that come with signing and encourages signing plugins in the future.

scottrigby committed:
Change behavior for installing plugin with missing .prov file (now warns and continues instead of failing)

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby commented:
@sabre1041 ok, pushed 0d54097 which updates the behavior for installing a plugin with missing .prov file to now emit a warning and continue, instead of failing with an error as it previously did.

```console
% go run ./cmd/helm plugin install https://github.com/scottrigby/h4-example-plugins/raw/refs/heads/main/tarballs/example-unsigned-0.1.0.tgz               
Verifying plugin signature...
WARNING: No provenance file found for plugin. Plugin is not signed and cannot be verified.
Installed plugin: example-unsigned
```

adding missing unit tests to check this, coming next.

scottrigby committed:
Add comprehensive plugin verification tests

- Test missing .prov files (warns but continues)
- Test invalid .prov file formats (fails verification)
- Test hash mismatches in .prov files (fails verification)
- Test .prov file access errors (fails appropriately)
- Test directory plugins don't support verification
- Test installation without verification enabled (succeeds)
- Test with valid .prov files (fails on empty keyring as expected)

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby commented:
@andy - i just added comprehensive plugin verification tests in 2adb5a890:
- Test missing .prov files (warns but continues)
- Test invalid .prov file formats (fails verification)
- Test hash mismatches in .prov files (fails verification)
- Test .prov file access errors (fails appropriately)
- Test directory plugins don't support verification
- Test installation without verification enabled (succeeds)
- Test with valid .prov files (fails on empty keyring as expected)

scottrigby committed:
Add missing cmd/plugin_verify CLI tests

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby commented:
also added remaining missing test coverage for this PR (cmd/plugin_verify CLI tests) in 5746baef3

gjenkins8 commented on internal/plugin/sign.go:
```diff
@@ -0,0 +1,166 @@
+}
+
+// ExtractPluginMetadataFromReader extracts plugin metadata from a tarball reader
+func ExtractPluginMetadataFromReader(r io.Reader) (*Metadata, error) {
```
```suggestion
func ExtractTgzPluginMetadata(r io.Reader) (*Metadata, error) {
```

scottrigby replied:
cool to do this one in this PR?
- https://github.com/helm/helm/pull/31196

scottrigby replied:
added to #31196 :+1: i agree it's a more explicit name

scottrigby commented on internal/plugin/installer/http_installer_test.go:
```diff
@@ -49,7 +49,7 @@ func (t *TestHTTPGetter) Get(_ string, _ ...getter.Option) (*bytes.Buffer, error
 
 // Fake plugin tarball data
-var fakePluginB64 = "H4sIAKRj51kAA+3UX0vCUBgGcC9jn+Iwuk3Peza3GeyiUlJQkcogCOzgli7dJm4TvYk+a5+k479UqquUCJ/fLs549sLO2TnvWnJa9aXnjwujYdYLovxMhsPcfnHOLdNkOXthM/IVQQYjg2yyLLJ4kXGhLp5j0z3P41tZksqxmspL3B/O+j/XtZu1y8rdYzkOZRCxduKPk53ny6Wwz/GfIIf1As8lxzGJSmoHNLJZphKHG4YpTCE0wVk3DULfpSJ3DMMqkj3P5JfMYLdX1Vr9Ie/5E5cstcdC8K04iGLX5HaJuKpWL17F0TCIBi5pf/0pjtLhun5j3f9v6r7wfnI/H0eNp9d1/5P6Gez0vzo7wsoxfrAZbTny/o9k6J ...
+var fakePluginB64 = "H4sIAAAAAAAAA+3SQUvDMBgG4Jz7K0LwapdvSxrwJig6mCKC5xHabBaXdDSt4L+3cQ56mV42ZPg+lw+SF5LwZmXf3OV206/rMGEnIgdG6zTJaDmee4y01FOlZpqGHJGZSsb1qS401sfOtpyz0FTup9xv+2dqNep/N/IP6zdHPSMVXCh1sH8yhtGMDBUFFTL1r4iIcXnUWxzwz/sP1rsrLkbfQGTvro11E4ZlmcucRNZHu04py1OO73OVi2Vbb7td9vp7nXevtvsKRpGVjfc2VMP2xf3t4mH5tHi5mz8ub+bPk9JXIvvr5wMAAAAAAAAAAAAAAAAAAAAAnLVPqwHcXQAoAAA="
```
not specifically, but we support legacy plugins so it worked. can we change this in future when needed?

scottrigby committed:
Make extractPluginMetadata private

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby commented on internal/plugin/sign.go:
```diff
@@ -0,0 +1,166 @@
+}
+
+// ExtractPluginMetadata extracts plugin metadata from a tarball
+func ExtractPluginMetadata(tarballPath string) (*Metadata, error) {
```
done in c8086d08d
